File: packages/backend/src/setup/prompt.ts
import { createInterface } from 'node:readline/promises'
import * as v from 'valibot'

/**
 * The terminal half of the `setup` command: one question at a time, on stdin and stdout.
 *
 * Three shapes of question and nothing else — pick one of a few, type a value, type a value
 * nobody should see. There is no prompt library behind them: `AGENTS.md` section 14 asks
 * what a dependency is worth, and a dependency that runs once per installation, in a command
 * that is already the only place in the codebase reading a terminal, is not worth a
 * lockfile entry — and the counterpart in `fast-servers/backend/src/prompt.c` has to be
 * written by hand either way, so this is the shape both of them keep.
 *
 * **Every question has a default and Enter takes it.** A setup that can be finished by
 * holding Enter down is a setup somebody will actually run; the defaults are what the
 * conversation in `askForSetup.ts` proposes, and what it proposes is what the environment
 * already says.
 */

/** Whether there is somebody at the other end to answer. */
export function isTerminal(): boolean {
  return process.stdin.isTTY === true && process.stdout.isTTY === true
}

export type Choice<T> = {
  readonly value: T
  readonly label: string
  /** One clause after the label, saying what choosing it means. */
  readonly hint?: string
}

/**
 * One of a handful, chosen with the arrow keys and Enter.
 *
 * Raw mode is what makes a cursor possible, and a terminal that will not give it — a pipe,
 * a CI job, a Windows console this build cannot switch — gets the same question as a
 * numbered list to type an answer to. The choice is the same either way; only the typing is.
 */
export async function askChoice<T>(
  question: string,
  choices: readonly Choice<T>[],
  initial = 0,
): Promise<T> {
  process.stdout.write(`\n${question}\n`)
  const stdin = process.stdin
  if (!isTerminal() || typeof stdin.setRawMode !== 'function') {
    return askNumbered(choices, initial)
  }

  return new Promise<T>((resolve) => {
    let index = initial
    stdin.setRawMode(true)
    stdin.resume()
    render(choices, index)

    let answered = false

    const finish = (chosen: number): void => {
      answered = true
      stdin.off('data', onData)
      stdin.setRawMode(false)
      stdin.pause()
      /* The list collapses to the answer, so that a finished setup reads back as questions
         and answers rather than as a screenful of menus. */
      process.stdout.write(`\u001b[${choices.length}A\u001b[J❯ ${choices[chosen]?.label ?? ''}\n`)
      resolve(choices[chosen]?.value as T)
    }

    const onData = (chunk: Buffer): void => {
      for (const key of keysIn(chunk)) {
        if (answered) {
          return
        }
        if (key === '\u0003') {
          /* Ctrl-C, which raw mode does not turn into a signal. 130 is what a shell reports
             for a program that ended on SIGINT, and setup has written nothing at this point. */
          stdin.setRawMode(false)
          process.stdout.write('\n')
          process.exit(130)
        }
        if (key === '\r' || key === '\n') {
          finish(index)
          continue
        }
        const digit = Number.parseInt(key, 10)
        if (Number.isInteger(digit) && digit >= 1 && digit <= choices.length) {
          finish(digit - 1)
          continue
        }
        const moved = step(key)
        if (moved !== 0) {
          index = (index + moved + choices.length) % choices.length
          process.stdout.write(`\u001b[${choices.length}A`)
          render(choices, index)
        }
      }
    }

    stdin.on('data', onData)
  })
}

/** Yes or no, as a choice — the same cursor, and Enter on what is already proposed. */
export async function askYesNo(question: string, initial: boolean): Promise<boolean> {
  return askChoice<boolean>(
    question,
    [
      { value: true, label: 'yes' },
      { value: false, label: 'no' },
    ],
    initial ? 0 : 1,
  )
}

/**
 * One typed value, checked against the same schema the rest of the system checks it with.
 *
 * An empty line takes `fallback`, which is what the parentheses after the label show. A
 * rejected answer is asked again rather than ending the setup: somebody is standing at the
 * terminal, and losing eleven correct answers because the twelfth had a typo would be
 * gratuitous.
 */
export async function askText<TSchema extends v.GenericSchema<string, unknown>>(
  label: string,
  fallback: string,
  schema: TSchema,
): Promise<v.InferOutput<TSchema>> {
  for (;;) {
    const typed = await readLine(fallback === '' ? `${label}: ` : `${label} (${fallback}): `)
    const answer = v.safeParse(schema, typed === '' ? fallback : typed)
    if (answer.success) {
      return answer.output
    }
    /* The schema's own message, which is the one the admin frontend will show for the same
       field later. Written to stderr so a piped stdout stays clean. */
    process.stderr.write(`  ${answer.issues[0].message}\n`)
  }
}

/**
 * A value that must not end up in the scrollback of a shared terminal.
 *
 * Echoed as asterisks, which needs raw mode; without it the answer is read the ordinary way
 * and is visible, because a setup that cannot ask for a password at all is worse than one
 * that asks for it in the open. An empty line keeps whatever is configured now, so a rerun
 * of `setup` does not have to retype it.
 */
export async function askSecret(label: string, fallback: string): Promise<string> {
  const shown = fallback === '' ? 'none' : 'unchanged'
  const stdin = process.stdin
  if (!isTerminal() || typeof stdin.setRawMode !== 'function') {
    const typed = await readLine(`${label} (${shown}): `)
    return typed === '' ? fallback : typed
  }

  process.stdout.write(`${label} (${shown}): `)
  return new Promise<string>((resolve) => {
    let typed = ''
    stdin.setRawMode(true)
    stdin.resume()

    let answered = false

    const onData = (chunk: Buffer): void => {
      for (const key of keysIn(chunk)) {
        if (answered) {
          return
        }
        if (key === '\u0003') {
          stdin.setRawMode(false)
          process.stdout.write('\n')
          process.exit(130)
        }
        if (key === '\r' || key === '\n') {
          answered = true
          stdin.off('data', onData)
          stdin.setRawMode(false)
          stdin.pause()
          process.stdout.write('\n')
          resolve(typed === '' ? fallback : typed)
          return
        }
        if (key === '\u007f' || key === '\b') {
          if (typed !== '') {
            typed = typed.slice(0, -1)
            process.stdout.write('\b \b')
          }
          continue
        }
        /* Everything else is text, except the escape sequences an arrow key sends — there is
           nothing to move a cursor over here, and letting them through would put a `[D` in a
           password. */
        if (key.startsWith('\u001b')) {
          continue
        }
        typed += key
        process.stdout.write('*')
      }
    }

    stdin.on('data', onData)
  })
}

/** A sentence that is part of the conversation rather than part of the log. */
export function say(...lines: readonly string[]): void {
  process.stdout.write(`${lines.join('\n')}\n`)
}

async function readLine(query: string): Promise<string> {
  /* One interface per question, because `askChoice` and `askSecret` read the same stdin
     directly and two readers on one stream take a keystroke each. */
  const io = createInterface({ input: process.stdin, output: process.stdout })
  try {
    return (await io.question(query)).trim()
  } finally {
    io.close()
  }
}

/**
 * One chunk from a terminal, as the keys that are in it.
 *
 * A keystroke is usually a chunk, and is not always: a paste arrives as one, and so does an
 * arrow key pressed a moment before Enter — the terminal hands over whatever has accumulated.
 * A handler that compared the whole chunk to `\r` would ignore both.
 */
function keysIn(chunk: Buffer): string[] {
  const text = chunk.toString('utf8')
  const keys: string[] = []
  for (let at = 0; at < text.length; ) {
    /* A CSI sequence — what every arrow key sends — is three characters and one key. */
    const csi = text.startsWith('\u001b[', at)
    keys.push(text.slice(at, at + (csi ? 3 : 1)))
    at += csi ? 3 : 1
  }
  return keys
}

/** Which way an arrow key points, or 0 for a key that is not one. */
function step(key: string): number {
  if (key === '\u001b[A' || key === '\u001b[D') {
    return -1
  }
  if (key === '\u001b[B' || key === '\u001b[C') {
    return 1
  }
  return 0
}

function render<T>(choices: readonly Choice<T>[], index: number): void {
  for (const [at, choice] of choices.entries()) {
    const chosen = at === index
    const label = chosen ? `\u001b[1m${choice.label}\u001b[22m` : choice.label
    const hint = choice.hint === undefined ? '' : `  — ${choice.hint}`
    /* Cleared line by line rather than as a block: the line before may have been longer, and
       half a hint left standing under a shorter one reads as part of it. */
    process.stdout.write(`\u001b[2K${chosen ? '❯ ' : '  '}${label}${hint}\n`)
  }
}

async function askNumbered<T>(choices: readonly Choice<T>[], initial: number): Promise<T> {
  for (const [at, choice] of choices.entries()) {
    const hint = choice.hint === undefined ? '' : ` — ${choice.hint}`
    process.stdout.write(`  ${at + 1}) ${choice.label}${hint}\n`)
  }
  for (;;) {
    const typed = await readLine(`Choose 1-${choices.length} (${initial + 1}): `)
    if (typed === '') {
      return choices[initial]?.value as T
    }
    const chosen = choices.findIndex(
      (choice, at) =>
        typed === String(at + 1) || typed.toLowerCase() === choice.label.toLowerCase(),
    )
    if (chosen !== -1) {
      return choices[chosen]?.value as T
    }
    process.stderr.write(`  Please answer with a number between 1 and ${choices.length}\n`)
  }
}

File: fast-servers/backend/src/prompt.c
/*
 * The setup command's questions. prompt.h is the specification and
 * packages/backend/src/setup/prompt.ts is the reference this is held to.
 */
#include "prompt.h"

#include <signal.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#if defined(_WIN32)
#include <io.h>
#define bk_isatty(stream) _isatty(_fileno(stream))
#else
#include <sys/select.h>
#include <termios.h>
#include <unistd.h>
#define bk_isatty(stream) isatty(fileno(stream))
#endif

/** What a keystroke means here. Everything else is a character or is ignored. */
enum {
    KEY_EOF = -1,
    KEY_ENTER = -2,
    KEY_UP = -3,
    KEY_DOWN = -4,
    KEY_BACKSPACE = -5,
    KEY_INTERRUPT = -6,
    KEY_IGNORED = -7
};

int bk_prompt_is_terminal(void)
{
    return bk_isatty(stdin) && bk_isatty(stdout);
}

void bk_say(const char *fmt, ...)
{
    va_list args;

    va_start(args, fmt);
    (void)vfprintf(stdout, fmt, args);
    va_end(args);
    (void)fputc('\n', stdout);
    (void)fflush(stdout);
}

#if defined(_WIN32)

/* No raw mode here: the console's own API is a different thing from termios, and a cursor
 * nobody on this platform runs a server on is not worth a second implementation. Everything
 * below falls back to the typed form, which is the same fallback the reference path takes
 * when it cannot switch a terminal to raw mode. */
static int raw_begin(void)
{
    return 0;
}
static void raw_end(void)
{
}
static int read_key(void)
{
    return KEY_EOF;
}

/* There is no terminal state of ours to put back, so ending is all there is to do. */
static void on_interrupt(int signal_number)
{
    (void)signal_number;
    _exit(130);
}

#else

static struct termios g_saved;

/** Whether the terminal is in raw mode now -- what on_interrupt reads to know that it has
 *  something to put back. */
static volatile sig_atomic_t g_raw;

/**
 * Turns off echo and line buffering, so that a keystroke arrives as it is pressed -- and turns
 * off the signal keys, so that Ctrl-C arrives as a keystroke too, byte 3, and is answered where
 * the terminal is being read rather than in a handler. That is what raw mode means to Node's
 * setRawMode, which the reference path uses; without ISIG off, Ctrl-C would still be a SIGINT,
 * and the handler the process has for it while serving only raises a flag nobody here reads.
 */
static int raw_begin(void)
{
    struct termios raw;

    if (tcgetattr(fileno(stdin), &g_saved) != 0)
        return 0;
    raw = g_saved;
    raw.c_lflag &= (tcflag_t) ~(ICANON | ECHO | ISIG);
    raw.c_cc[VMIN] = 1;
    raw.c_cc[VTIME] = 0;
    if (tcsetattr(fileno(stdin), TCSANOW, &raw) != 0)
        return 0;
    g_raw = 1;
    return 1;
}

static void raw_end(void)
{
    g_raw = 0;
    (void)tcsetattr(fileno(stdin), TCSANOW, &g_saved);
}

/**
 * SIGINT while the conversation is held -- see bk_prompt_begin().
 *
 * Only what may be done in a handler: tcsetattr, write and _exit are all async-signal-safe.
 * Raw mode is put back for the case where a signal arrives anyway -- somebody sending one with
 * kill rather than with the keyboard -- because a shell left without echo is worse than any
 * setup that did not finish.
 */
static void on_interrupt(int signal_number)
{
    ssize_t written;

    (void)signal_number;
    if (g_raw)
        (void)tcsetattr(STDIN_FILENO, TCSANOW, &g_saved);
    written = write(STDOUT_FILENO, "\n", 1);
    (void)written;
    _exit(130);
}

/** Whether another byte of the same escape sequence is already there. */
static int more_within(long micros)
{
    fd_set set;
    struct timeval wait;

    FD_ZERO(&set);
    FD_SET(fileno(stdin), &set);
    wait.tv_sec = 0;
    wait.tv_usec = micros;
    return select(fileno(stdin) + 1, &set, NULL, NULL, &wait) > 0;
}

static int read_key(void)
{
    unsigned char c;

    if (read(fileno(stdin), &c, 1) != 1)
        return KEY_EOF;
    if (c == 3)
        return KEY_INTERRUPT;
    if (c == '\r' || c == '\n')
        return KEY_ENTER;
    if (c == 127 || c == 8)
        return KEY_BACKSPACE;
    if (c != 27)
        return (int)c;

    /* An escape, which is either an arrow key or somebody pressing Escape. The rest of the
     * sequence is already in the buffer when it is a key; waiting a moment for it is what
     * tells the two apart without a timer of our own. */
    if (!more_within(50000) || read(fileno(stdin), &c, 1) != 1 || c != '[')
        return KEY_IGNORED;
    if (!more_within(50000) || read(fileno(stdin), &c, 1) != 1)
        return KEY_IGNORED;
    if (c == 'A' || c == 'D')
        return KEY_UP;
    if (c == 'B' || c == 'C')
        return KEY_DOWN;
    return KEY_IGNORED;
}

#endif

/** What SIGINT did before bk_prompt_begin(), for bk_prompt_end() to put back. */
static void (*g_previous_interrupt)(int) = SIG_DFL;

void bk_prompt_begin(void)
{
    void (*previous)(int) = signal(SIGINT, on_interrupt);

    if (previous != SIG_ERR)
        g_previous_interrupt = previous;
}

void bk_prompt_end(void)
{
    (void)signal(SIGINT, g_previous_interrupt);
}

static void render(const char *const *labels, const char *const *hints, int count, int index)
{
    int i;

    for (i = 0; i != count; ++i) {
        const char *hint = hints != NULL && hints[i] != NULL ? hints[i] : NULL;

        /* Cleared line by line rather than as a block: the line before may have been longer,
         * and half a hint left standing under a shorter one reads as part of it. */
        (void)fprintf(stdout, "\033[2K%s%s%s%s%s%s\n", i == index ? "\xe2\x9d\xaf " : "  ",
                      i == index ? "\033[1m" : "", labels[i], i == index ? "\033[22m" : "",
                      hint != NULL ? "  \xe2\x80\x94 " : "", hint != NULL ? hint : "");
    }
    (void)fflush(stdout);
}

/** The list as something to type a number into, for a terminal that has no raw mode. */
static int choose_numbered(const char *const *labels, const char *const *hints, int count,
                           int initial)
{
    char line[BK_ANSWER_MAX];
    int i;

    for (i = 0; i != count; ++i) {
        const char *hint = hints != NULL && hints[i] != NULL ? hints[i] : NULL;

        (void)fprintf(stdout, "  %d) %s%s%s\n", i + 1, labels[i],
                      hint != NULL ? " \xe2\x80\x94 " : "", hint != NULL ? hint : "");
    }
    for (;;) {
        (void)fprintf(stdout, "Choose 1-%d (%d): ", count, initial + 1);
        (void)fflush(stdout);
        if (fgets(line, sizeof(line), stdin) == NULL)
            return -1;
        for (i = 0; line[i] != '\0'; ++i) {
            if (line[i] == '\n' || line[i] == '\r') {
                line[i] = '\0';
                break;
            }
        }
        if (line[0] == '\0')
            return initial;
        for (i = 0; i != count; ++i) {
            if ((line[0] == (char)('1' + i) && line[1] == '\0') || strcmp(line, labels[i]) == 0)
                return i;
        }
        (void)fprintf(stderr, "  Please answer with a number between 1 and %d\n", count);
    }
}

int bk_prompt_choose(const char *question, const char *const *labels, const char *const *hints,
                     int count, int initial)
{
    int index = initial;

    if (labels == NULL || count <= 0)
        return -1;
    if (initial < 0 || initial >= count)
        index = 0;

    (void)fprintf(stdout, "\n%s\n", question);
    (void)fflush(stdout);
    if (!bk_prompt_is_terminal() || !raw_begin())
        return choose_numbered(labels, hints, count, index);

    render(labels, hints, count, index);
    for (;;) {
        int key = read_key();

        if (key == KEY_EOF) {
            raw_end();
            return -1;
        }
        if (key == KEY_INTERRUPT) {
            /* Ctrl-C, which raw mode does not turn into a signal -- see raw_begin. 130 is what a
             * shell reports for a program that ended on SIGINT, and setup has written nothing at
             * this point. */
            raw_end();
            (void)fputc('\n', stdout);
            exit(130);
        }
        if (key == KEY_ENTER)
            break;
        if (key >= '1' && key < '1' + count) {
            index = key - '1';
            break;
        }
        if (key == KEY_UP || key == KEY_DOWN) {
            index = (index + (key == KEY_UP ? count - 1 : 1)) % count;
            (void)fprintf(stdout, "\033[%dA", count);
            render(labels, hints, count, index);
        }
    }

    raw_end();
    /* The list collapses to the answer, so that a finished setup reads back as questions and
     * answers rather than as a screenful of menus. */
    (void)fprintf(stdout, "\033[%dA\033[J\xe2\x9d\xaf %s\n", count, labels[index]);
    (void)fflush(stdout);
    return index;
}

int bk_prompt_yes_no(const char *question, int initial)
{
    static const char *const kLabels[] = {"yes", "no"};
    int chosen = bk_prompt_choose(question, kLabels, NULL, 2, initial ? 0 : 1);

    return chosen < 0 ? -1 : (chosen == 0);
}

/** Copies @p value into @p out, or answers 0 when it does not fit. */
static int fits(const char *value, char *out, size_t out_size)
{
    size_t len = strlen(value);

    if (len + 1 > out_size)
        return 0;
    memcpy(out, value, len + 1);
    return 1;
}

int bk_prompt_text(const char *label, const char *fallback, char *out, size_t out_size)
{
    char line[BK_ANSWER_MAX];

    if (fallback == NULL)
        fallback = "";
    for (;;) {
        size_t length;
        size_t begin = 0;

        if (fallback[0] == '\0')
            (void)fprintf(stdout, "%s: ", label);
        else
            (void)fprintf(stdout, "%s (%s): ", label, fallback);
        (void)fflush(stdout);
        if (fgets(line, sizeof(line), stdin) == NULL)
            return 0;

        length = strlen(line);
        while (length != 0 && (line[length - 1] == '\n' || line[length - 1] == '\r' ||
                               line[length - 1] == ' ' || line[length - 1] == '\t'))
            line[--length] = '\0';
        while (begin != length && (line[begin] == ' ' || line[begin] == '\t'))
            ++begin;

        if (fits(begin == length ? fallback : line + begin, out, out_size))
            return 1;
        (void)fprintf(stderr, "  This value is too long\n");
    }
}

int bk_prompt_secret(const char *label, const char *fallback, const char *shown, char *out,
                     size_t out_size)
{
    char typed[BK_ANSWER_MAX];

    if (fallback == NULL)
        fallback = "";
    if (shown == NULL)
        shown = fallback[0] == '\0' ? "none" : "unchanged";
    for (;;) {
        size_t length = 0;

        (void)fprintf(stdout, "%s (%s): ", label, shown);
        (void)fflush(stdout);

        if (!bk_prompt_is_terminal() || !raw_begin()) {
            /* Visible, and the header says why that is the right trade against not asking at
             * all. The label has already been written, so this only reads. */
            if (fgets(typed, sizeof(typed), stdin) == NULL)
                return 0;
            length = strlen(typed);
            while (length != 0 && (typed[length - 1] == '\n' || typed[length - 1] == '\r'))
                typed[--length] = '\0';
        } else {
            for (;;) {
                int key = read_key();

                if (key == KEY_EOF) {
                    raw_end();
                    return 0;
                }
                if (key == KEY_INTERRUPT) {
                    raw_end();
                    (void)fputc('\n', stdout);
                    exit(130);
                }
                if (key == KEY_ENTER)
                    break;
                if (key == KEY_BACKSPACE) {
                    if (length != 0) {
                        --length;
                        (void)fputs("\b \b", stdout);
                        (void)fflush(stdout);
                    }
                    continue;
                }
                /* Anything that is not a character -- an arrow key among them -- is nothing to
                 * do here: there is no cursor to move over a value that is not shown. */
                if (key < 0)
                    continue;
                if (length + 1 < sizeof(typed)) {
                    typed[length++] = (char)key;
                    (void)fputc('*', stdout);
                    (void)fflush(stdout);
                }
            }
            raw_end();
            (void)fputc('\n', stdout);
            (void)fflush(stdout);
            typed[length] = '\0';
        }

        if (fits(length == 0 ? fallback : typed, out, out_size))
            return 1;
        (void)fprintf(stderr, "  This value is too long\n");
    }
}
